test/model/recent_senders_test.dart

```diff
@@ -0,0 +1,201 @@
+import 'package:checks/checks.dart';
+import 'package:test/scaffolding.dart';
+import 'package:zulip/api/model/events.dart';
+import 'package:zulip/api/model/model.dart';
+import 'package:zulip/model/recent_senders.dart';
+import '../example_data.dart' as eg;
+import 'recent_senders_checks.dart';
+
+void checkMatchesMessages(List<Message> messages, RecentSenders model) {
+  final Map<int, Map<int, MessageIdTracker>> expectedStreamSenders = {};
+  final Map<int, Map<String, Map<int, MessageIdTracker>>> expectedTopicSenders = {};
+  for (final message in messages) {
+    if (message is! StreamMessage) continue;
+
+    final StreamMessage(:streamId, :topic, :senderId, id: int messageId) = message;
+
+    // Track message in stream.
+    var sendersMap = expectedStreamSenders[streamId] ??= {};
+    var idTracker = sendersMap[senderId] ??= MessageIdTracker();
+    idTracker.add(messageId);
+
+    // Track message in topic.
+    final topicsMap = expectedTopicSenders[streamId] ??= {};
+    sendersMap = topicsMap[topic] ??= {};
+    idTracker = sendersMap[senderId] ??= MessageIdTracker();
+    idTracker.add(messageId);
+  }
+  check(model)
+    ..streamSenders.deepEquals(expectedStreamSenders)
+    ..topicSenders.deepEquals(expectedTopicSenders);
+}
+
+void main() {
+  group('RecentSenders', () {
+    test('starts with empty stream and topic senders', () {
+      final model = RecentSenders();
+      checkMatchesMessages([], model);
+    });
+
+    test('just process a stream message', () {
```

Comment:
This is clearest as two cases:
* stream message gets included
* DM gets ignored